File: MainApp/src/main/java/com/sleticalboy/dailywork/MainApp.java
package com.sleticalboy.dailywork;

import android.app.Application;
import android.content.Context;

import com.sleticalboy.dailywork.bean.DaoMaster;
import com.sleticalboy.dailywork.bean.DaoSession;
import com.sleticalboy.dailywork.cache.ACache;

import org.greenrobot.greendao.database.Database;

import java.lang.ref.Reference;
import java.lang.ref.WeakReference;

/**
 * Created on 18-3-5.
 *
 * @author sleticalboy
 * @version 1.0
 * @description
 */
public class MainApp extends Application {

    private static DaoSession sDaoSession;
    private static Reference<Context> mRefContext;

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        mRefContext = new WeakReference<>(base);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        initCache();
        initDB();
    }

    private void initCache() {
        ACache.mCacheName = "DailyWork";
    }

    private void initDB() {
        DaoMaster.DevOpenHelper helper = new DaoMaster.DevOpenHelper(this, "daily-work");
        Database database = helper.getWritableDb();
        sDaoSession = new DaoMaster(database).newSession();
    }

    public static DaoSession getDaoSession() {
        return sDaoSession;
    }

    public static Context getContext() {
        return mRefContext.get();
    }
}

File: MainApp/src/main/java/com/sleticalboy/dailywork/MainApp.kt
package com.sleticalboy.dailywork

import android.app.Application
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import android.util.Log
import com.sleticalboy.dailywork.components.receiver.GlobalReceiver
import com.sleticalboy.util.NotificationHelper
import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher

/**
 * Created on 18-3-5.
 *
 * @author leebin
 * @version 1.0
 */
class MainApp : Application() {

    override fun onCreate() {
        super.onCreate()
        mApp = this
        installLeakCanary()
        adaptAndroidO()
    }

    private fun installLeakCanary() {
        if (!LeakCanary.isInAnalyzerProcess(this)) {
            mRefWatcher = LeakCanary.install(this)
        }
    }

    private fun adaptAndroidO() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            return
        }
        Log.d("MainApp", "adapt android O")
        registerReceivers()
        NotificationHelper.createAllChannels(this)
    }

    private fun registerReceivers() {
        val intentFilter = IntentFilter()
        intentFilter.addAction(Intent.ACTION_SCREEN_ON)
        intentFilter.addAction(Intent.ACTION_USER_PRESENT)
        registerReceiver(GlobalReceiver(), intentFilter)
    }

    companion object {

        private var mApp: Application? = null
        private var mRefWatcher: RefWatcher? = null

        val app: Context? get() = mApp
    }
}
